File: call-notes-to-pedigree/node/src/transcriptSource.ts
/**
 * Read a transcript from a file path or from a stdin-like stream.
 */

import { readFile } from 'node:fs/promises';
import type { Readable } from 'node:stream';

export class TranscriptError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TranscriptError';
  }
}

export async function readFromPath(path: string): Promise<string> {
  let text: string;
  try {
    text = await readFile(path, 'utf8');
  } catch (error) {
    const reason = error instanceof Error ? error.message : String(error);
    throw new TranscriptError(`Could not read transcript ${path}: ${reason}`);
  }
  return requireNonEmpty(text, path);
}

export async function readFromStream(stream: Readable): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of stream) {
    chunks.push(typeof chunk === 'string' ? Buffer.from(chunk) : (chunk as Buffer));
  }
  return requireNonEmpty(Buffer.concat(chunks).toString('utf8'), 'stdin');
}

function requireNonEmpty(text: string, source: string): string {
  if (text.trim() === '') {
    throw new TranscriptError(`Transcript from ${source} was empty.`);
  }
  return text;
}

File: call-notes-to-pedigree/python/src/call_notes_to_pedigree/transcript_source.py
"""Read a transcript from a file path or from a stdin-like stream."""

from __future__ import annotations

from pathlib import Path
from typing import TextIO


class TranscriptError(ValueError):
    """Raised when no usable transcript can be read."""


def read_from_path(path: Path) -> str:
    try:
        text = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise TranscriptError(f"Could not read transcript {path}: {exc}") from exc
    return _require_non_empty(text, source=str(path))


def read_from_stream(stream: TextIO) -> str:
    return _require_non_empty(stream.read(), source="stdin")


def _require_non_empty(text: str, *, source: str) -> str:
    if not text.strip():
        raise TranscriptError(f"Transcript from {source} was empty.")
    return text
